### PR TITLE
Learn

### PR DESCRIPTION
**Pull request by rtibbles**
## Summary

A Skeleton Learn App to demonstrate plugin functionality.

## TODO

- [ ] Have tests been written for the new code?
- [ ] Has documentation been written/updated?
- [ ] New dependencies (if any) added to requirements file
- [ ] Add an entry to CHANGELOG.rst
- [ ] Add yourself it AUTHORS.rst if you don't appear there

## Reviewer guidance

Build frontend assets, run Kolibri and click through to 'Learn' from the menu.

## Documentation

None yet

### CODE REVIEW

**Comment by rtibbles** [comment]
@indirectlylit @MCGallaspy @benjaoming @jamalex @aronasorman @DXCanas

Please give feedback and thoughts.

The first commit at least should be merged at some point due to the general plugin functionality it exposes, the second commit might be useful boilerplate for the learn plugin.

If it is useful to merge, I will add tests at a later date.

**Comment by MCGallaspy** [comment]
Can you summarize the changes?

File: kolibri/core/templates/kolibri/base.html [hunk 1]
```diff
@@ -2,9 +2,11 @@
 <html>
 <head>
   <title>{% block title %}{% endblock %} - {% trans "Kolibri" %}</title>
+  {% block frontend_assets %}
```
Comment (rtibbles) [inline kolibri/core/templates/kolibri/base.html:5]: Add a block tag to allow injection of frontend_assets in derived templates.

File: kolibri/core/urls.py [hunk 2]
```diff
@@ -4,12 +4,22 @@
...
 from django.conf.urls import url
+from kolibri.plugins import hooks
 
 from . import views
 
 app_name = 'kolibri'
 
 urlpatterns = [
-    url('^', views.IndexView.as_view(), name='index'),
+    url('^/$', views.IndexView.as_view(), name='index'),
```
Comment (rtibbles) [inline kolibri/core/urls.py:14]: Prevent the base URL catching _all_ URLs.

File: kolibri/core/urls.py [hunk 3]
```diff
@@ -4,12 +4,22 @@
...
     # url('.* ', views.TODOView.as_view())
 ]
+
+def retrieve_plugin_urls():
```
Comment (rtibbles) [inline kolibri/core/urls.py:18]: Get all the URLs registered by plugins against the IMPORT_URLS hook - the functions registered should return the arguments for the django `url` function in a dict with `url_base` and `urls`.

File: kolibri/plugins/assets/src/kolibri_module/kolibri_module.js [hunk 4]
```diff
@@ -58,6 +58,15 @@ KolibriModule.prototype.stop_listening = function(event, method) {
     Kolibri.stop_listening(event, this, method);
 };
 
+/**
+ * Convenience method to fire an event on the global Mediator and pass in a data payload.
+ * @param {string} event - the event name
+ * @param [arguments] arguments - additional arguments to the event handler.
+ */
+KolibriModule.prototype.trigger = function() {
```
Comment (rtibbles) [inline kolibri/plugins/assets/src/kolibri_module/kolibri_module.js:66]: This allows any plugin to easily trigger an event on the Mediator.

File: kolibri/plugins/hooks.py [hunk 5]
```diff
@@ -35,6 +35,8 @@ def __enumerate_hook():
 FRONTEND_PLUGINS = __enumerate_hook()
 BASE_FRONTEND_SYNC = __enumerate_hook()
 BASE_FRONTEND_ASYNC = __enumerate_hook()
+# Add a hook for namespaced URL importing from plugins
+IMPORT_URLS = __enumerate_hook()
```
Comment (rtibbles) [inline kolibri/plugins/hooks.py:41]: Add a hook for adding plugin URLs to Kolibri.

**Comment by rtibbles** [comment]
In Summary this PR does two things:

1) Adds features that help with plugins and modules such as:
* Provides a mechanism for Plugins to register their own URLs modules to the Core (appropriately namespaced).
* Provides a trigger method on the Base `KolibriModule` class so that KolibriModules can easily trigger events on the Mediator.
* Adds a frontend asset block tag to the base.html to allow Django templates that extend it can add more frontend assets.

2) Provides a skeleton Learn app with the following features:
* Shows three buttons representing content nodes of different types.
* Each content node button triggers a different kind of 'content_render' event specific to a content mimetype.
* When the button is pressed, the appropriate event is triggered.
* In addition, three content renderer plugins are defined, each of which listens to a specific mimetype.
* These plugins are all asynchronously registered, so are not in the page at page load.
* Once a button has been pressed, the appropriate plugin receives data about the event, and a reference for where to render its content.
* The plugin renders its content into the element provided.

What it does not do:
* Clear up the DOM, it just keeps appending nodes.
* Do any kind of URL routing, so it is unaware if it is already showing the current content node.
* Interface with the content app at all to actually receive metadata about content, these are just hardcoded.

File: kolibri/core/urls.py (hunk 3 above)
Comment (rtibbles) [inline kolibri/core/urls.py:18]: @indirectlylit This is where the urls defined in a plugin get inserted into the Kolibri Core `urls.py`.

File: kolibri/core/urls.py (hunk 2 above)
Comment (MCGallaspy) [inline kolibri/core/urls.py:14]: Probably not what you intended -- e.g. 404s on `http://127.0.0.1:8000/` but catches `http://127.0.0.1:8000//`. See the [APPEND_SLASH setting](https://docs.djangoproject.com/en/1.9/ref/settings/#append-slash).

**Comment by MCGallaspy** [comment]
Plugin must be enabled: `kolibri plugin kolibri.plugins.learn enable`.

**Comment by MCGallaspy** [comment]
In fact they *all* have to be enabled:
```
>> kolibri plugin kolibri.plugins.document_pdf_render enable
>> kolibri plugin kolibri.plugins.audio_mp4_render enable
>> kolibri plugin kolibri.plugins.video_mp4_render enable 
```

**Comment by MCGallaspy** [comment]
I think this LOC is the cause of an error for me: https://github.com/rtibbles/kolibri/blob/learn/kolibri/core/templatetags/kolibri_tags.py#L64

Looks like a fs path is being used directly as a url, which is an error with Windows paths.

**Comment by MCGallaspy** [comment]
After looking at L64 and peering at the tests for a bit, I don't understand why you would want to use the fs path as a url at all, when it seems like you always add a url attribute to the dictionary in question anyway: https://github.com/rtibbles/kolibri/blob/learn/kolibri/utils/webpack.py#L119

Note the LOC just above 119 also uses non-Windows friendly paths, although Django seems to handle it ok.

I think there are two possibilities:
1. Don't use the fs path at all to build urls in template tags. This is preferable, unless you have some edge case you're thinking of? But if I modify `kolibri_tags.render_as_url` to *just* use the url attribute instead of the publicPath, everything *seems* to work OK (tests pass, no obvious errors in the site).
2. Account for this previously unanticipated edge case. Means that `kolibri_tags.render_as_url` needs to anticipate being handed a non-Unix path.

For (2), I wrote a test case that looks something like this, which fails on my Windows machines but I expect will pass on Unix machines. In practice we will want test different hard-coded paths (not `__file__` so that the tests are not OS-dependent:

```py
    @override_settings(STATIC_URL=test_url)
    def test_render_as_url_public_path_generates_valid_url(self):
        chunk = {
            "publicPath": os.path.abspath(__file__)  # This should be a legitimate file path.
        }
        val = URLValidator()
        val(kolibri_tags.render_as_url(chunk))  # Will raise an error for invalid urls
```

File: kolibri/plugins/audio_mp3_render/kolibri_plugin.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+
+class KolibriAudioMP3RenderFrontEnd(KolibriFrontEndPluginBase):
+    """
+    The base learn code for the learn page.
+    """
+    entry_file = "assets/src/render_audio_mp3_module.js"
+
+    events = {
+        "content_render:audio/mp3": "render"
```
Comment (MCGallaspy) [inline kolibri/plugins/audio_mp3_render/kolibri_plugin.py:22]: Ok, good to demo event registration. Have you considered appropriate behavior if multiple plugins register to this callback? Is there a use case for that?

File: kolibri/plugins/learn/assets/src/learn_module.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+var _ = require('lodash');
+var logging = require('loglevel');
+
+// This is aliased for your convenience!
+var KolibriModule = require('kolibri_module');
+
+var LearnModule = KolibriModule.extend({
+
+    start: function(plugin) {
+        // This is called on the kolibri_register event - make sure it is this plugin that has registered.
```
Comment (MCGallaspy) [inline kolibri/plugins/learn/assets/src/learn_module.js:12]: Would be good to document the built-in events that are triggered by the asset loader. Also seems like this may be redundant with the `initialize` function -- is there a difference?

File: kolibri/plugins/learn/templates/learn/learn.html [hunk 8]
```diff
@@ -0,0 +1,11 @@
+{% extends "kolibri/base.html" %}
+{% load learn_tags kolibri_tags %}
+{% block frontend_assets %}
+{{ block.super }}
+{% frontend_assets 'kolibri.plugins.learn.KolibriLearnFrontEnd' %}
+{% content_renderer_frontend_async %}
```
Comment (MCGallaspy) [inline kolibri/plugins/learn/templates/learn/learn.html:6]: With synchronous loading, I think it makes sense to have different tags, like `xyz_sync` for loading a set of resources and `frontend_assets` for loading an individual resource. But for async loading, why have multiple async tags? Do you think it makes sense to have _all_ modules included with e.g. the `base_async` tag? That way _any_ module could in principle be loaded async with no extra effort, whereas here I have to keep track of which modules are registered to `content_renderer_frontend_async` and also `base_async`, as well as manually registering my async modules through the plugin system.

File: kolibri/core/urls.py (hunk 2 above)
Comment (rtibbles) [inline kolibri/core/urls.py:14]: No, I should have just added the $.

**Comment by rtibbles** [comment]
Thanks for the catch - I think this was left over from the django module that I originally used (and then monkey patched, and then modified sufficiently to no longer import) to import webpack output into Django relatively seamlessly, can fix this up.

File: kolibri/plugins/audio_mp3_render/kolibri_plugin.py (hunk 6 above)
Comment (rtibbles) [inline kolibri/plugins/audio_mp3_render/kolibri_plugin.py:22]: I can't think of a use case for it immediately - but there may be a case where we might want to register `content_render:*` (or some other catch all - if we're going to stick with Backbone style events, then maybe triggering `content_render:audio/mp3` also automatically triggers `content_render`) - and then that will trigger the content recommendation to load relevant content that a learner might want to navigate to next.

In this case, I think it is fine to have multiple event listeners - and it seems like idiomatic Javascript too. What is less clear is if two different content playing plugins have both been registered at once. Do we want to legislate for that?

File: kolibri/plugins/learn/assets/src/learn_module.js (hunk 7 above)
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/learn_module.js:12]: Initialize is explicitly called _before_ registration with the Mediator - so initialize is only for things that you want to happen during the constructor function, before it is registered as a plugin.

Yes, I should list the event APIs of every KolibriModule, and the Core app. Will think on how to best present this.

File: kolibri/plugins/learn/templates/learn/learn.html (hunk 8 above)
Comment (rtibbles) [inline kolibri/plugins/learn/templates/learn/learn.html:6]: The purpose of the separate tags is to give a plugin developer a way to target their code. If it is only ever used on the learn tab I see no reason to have it loaded into every page, even if the amount of extra HTML it creates is minimal.

If there does become a use case for inserting it into all pages, then it can use that tag. I don't really want to default to overloading the base.html with everything, because I think it increases the risk of event collisions (someone specifies an event API in one KolibriModule, and it happens to have a name collision with another - we now have a KolibriModule responding to events it was not designed for.

So I guess my answer is - you should have to make an effort to load Javascript into a particular template, because that means you actually want it there.

File: kolibri/plugins/audio_mp3_render/kolibri_plugin.py (hunk 6 above)
Comment (MCGallaspy) [inline kolibri/plugins/audio_mp3_render/kolibri_plugin.py:22]: > What is less clear is if two different content playing plugins have both been registered at once. 

That's what I'm wondering about. If it doesn't make sense to have multiple plugins registered to handle the same content type then doing so should result in an error, maybe even at server start time. Otherwise the Learn page I think should gracefully handle this situation.

File: kolibri/plugins/learn/templates/learn/learn.html (hunk 8 above)
Comment (indirectlylit) [inline kolibri/plugins/learn/templates/learn/learn.html:6]: Do we expect that the most common scenario is that plugins should be loaded asynchronously?

File: kolibri/core/templates/kolibri/base.html (hunk 1 above)
Comment (rtibbles) [inline kolibri/core/templates/kolibri/base.html:5]: @MCGallaspy I think we should keep this.

File: kolibri/core/urls.py (hunk 3 above)
Comment (rtibbles) [inline kolibri/core/urls.py:18]: Happy to deprecate this in favour of whatever you and @benjaoming settle on, @MCGallaspy

File: kolibri/plugins/assets/src/kolibri_module/kolibri_module.js (hunk 4 above)
Comment (rtibbles) [inline kolibri/plugins/assets/src/kolibri_module/kolibri_module.js:66]: This is a useful method and should be kept, @MCGallaspy - we may want to refactor it based on @indirectlylit's conversion of Backbone to Vue, for event handling, however.

File: kolibri/plugins/audio_mp3_render/assets/src/render_audio_mp3_module.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+
+var _ = require('lodash');
+var logging = require('loglevel');
+
+// This is aliased for your convenience!
+var KolibriModule = require('kolibri_module');
+
+var AudioMP4Module = KolibriModule.extend({
+
+    render: function(content_data, render_area) {
```
Comment (rtibbles) [inline kolibri/plugins/audio_mp3_render/assets/src/render_audio_mp3_module.js:11]: I think most of these plugins should be kept, but the actual render methods can be removed for proper fill in with appropriate Vue logic later.

**Comment by rtibbles** [comment]
@MCGallaspy I think almost everything from here should be preserved, barring the URL handling, and the demo-specific native DOM rendering code in each of the KolibriModules defined in here.

**Comment by indirectlylit** [comment]
Commits from the PR are now in here: https://github.com/learningequality/kolibri/pull/70
